### PR TITLE
Deprecations and cleanup

### PR DESCRIPTION
**Pull request by TacoTheDank**
#### What is it?
- [ ] Bug fix (user facing)
- [ ] Feature (user facing)
- [x] Code base improvement (dev facing)
- [ ] Meta improvement to the project (dev facing)

#### Description of the changes in your PR

- Fixes a bunch of deprecations
- A little bit of code cleanup

#### Fixes the following issue(s)

- Fixes some deprecations in code

#### Testing apk

[newpipe-deprecationsandcleanup.zip](https://github.com/TeamNewPipe/NewPipe/files/4963478/newpipe-deprecationsandcleanup.zip)

#### Agreement
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

### CODE REVIEW

**Comment by wb9688** [comment]
Code looks OK, though have you tested your changes on at least Android 4.4 and Android 10?

We also prefer marking variables as `final` whenever possible, e.g. in those enhanced for loops.

**Comment by TacoTheDank** [comment]
@wb9688

I have no 4.4 device to test, and can't really do an AS emulator. I have an Android 10 one though, and will test on that.

I fixed a crash caused by one of the changes, so I reverted that (and better grouped the commits). I'm testing an APK, and everything works fine as far as I can tell.

**Comment by MD77MD** [comment]
I have a 4.4 device.. what should I look for to test... could you put the link to the APK.

**Comment by MD77MD** [comment]
I really think this will be more useful and needed after the big merge (expendable players with unified ui). with the release of newPipe 0.20.

**Comment by TacoTheDank** [comment]
@MD77MD Didn't see this. I've added the APK. (I think it's the right one)

**Comment by Stypox** [comment]
@MD77MD since you own a 4.4 device, are you willing to add yourself to the [testers page](https://github.com/TeamNewPipe/NewPipe/wiki/Testers) on the wiki?

**Comment by Domiuns** [comment]
I have test the last apk on my phone android 4.4 and all work well. I continue with test if i have issues i post here, thanks for work :)

**Comment by TobiGr** [comment]
Code looks good, but will be merged after #2907 as it changes player code.

**Comment by TacoTheDank** [comment]
Ok. I'll probably have to fix conflicts but that's fine.

**Comment by TacoTheDank** [comment]
Fixed. Some changes have been made as well, so re-review the commits.

**Comment by Redirion** [comment]
@TacoTheDank please update the ExoPlayer-Dependency to 2.11.8 as 2.11.8 contains only "[trivial changes](https://github.com/google/ExoPlayer/commit/185be99d6137d782aa5fb572cb7ef127456b0d26)" it should be safe to do so

File: app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ public Socket createSocket(final InetAddress address, final int port,
     }
 
     private Socket enableTLSOnSocket(final Socket socket) {
-        if (socket != null && (socket instanceof SSLSocket)) {
+        if ((socket instanceof SSLSocket)) {
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:109]: No need for the double `((` and `))`

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java [hunk 2]
```diff
@@ -572,7 +568,7 @@ private String trim(final String input) {
                 }
             }
         }
-        return result.toArray(new String[result.size()]);
+        return result.toArray(new String[0]);
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/RouterActivity.java:571]: Why was this changed?

File: app/src/main/java/org/schabi/newpipe/player/resolver/VideoPlaybackResolver.java [hunk 3]
```diff
@@ -115,7 +115,7 @@ public MediaSource resolve(@NonNull final StreamInfo info) {
             return mediaSources.get(0);
         } else {
             return new MergingMediaSource(mediaSources.toArray(
-                    new MediaSource[mediaSources.size()]));
+                    new MediaSource[0]));
```
Comment (wb9688) [inline app/src/main/java/org/schabi/newpipe/player/resolver/VideoPlaybackResolver.java:118]: Why was this changed?

**Review by wb9688 — CHANGES_REQUESTED** [review]
In general I like these changes. Could you rebase on `dev` and fix the Checkstyles errors though?

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java (hunk 2 above)
Comment (TacoTheDank) [reply app/src/main/java/org/schabi/newpipe/RouterActivity.java:571]: image

File: app/src/main/java/org/schabi/newpipe/player/resolver/VideoPlaybackResolver.java (hunk 3 above)
Comment (TacoTheDank) [reply app/src/main/java/org/schabi/newpipe/player/resolver/VideoPlaybackResolver.java:118]: See above.

File: app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java (hunk 1 above)
Comment (TacoTheDank) [reply app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:109]: Whoops.

File: app/src/main/java/org/schabi/newpipe/RouterActivity.java (hunk 2 above)
Comment (wb9688) [reply app/src/main/java/org/schabi/newpipe/RouterActivity.java:571]: Oh, OK

**Comment by wb9688** [comment]
You still haven't fixed all Checkstyle errors. Did you even try compiling the code before pushing it?

**Comment by TacoTheDank** [comment]
@wb9688 
> You still haven't fixed all Checkstyle errors. Did you even try compiling the code before pushing it?

Yeah, I did. I just had to undo the commits and re-push them. It's much easier to find where the errors are via travis than trying to play find-the-error with this:
image

Please watch the tone...

**Comment by wb9688** [comment]
@TacoTheDank: Ah, if it looks like that, it makes sense indeed, sorry. It shouldn't look like that though, are you using something other than Android Studio?

**Comment by TacoTheDank** [comment]
@wb9688 Nope, it's just plain ol' Android Studio. The runCheckstyle just tells me the number of errors/warnings in the files, but doesn't give me anything else to work with :/

**Comment by TacoTheDank** [comment]
I don't like how checkstyle operates. I've fixed all the errors, but it still errors because of the hashCode warning that I don't have anything to do with. I'll now have to fix that PR-unrelated thing for it to actually succeed :/

**Comment by wb9688** [comment]
@TacoTheDank: Oh, that's very strange. It should say e.g. the line number and that should be clickable to jump to the error. If you run `./gradlew runCheckstyle` in a terminal, does it show the errors and warnings with the line numbers (like on Travis)? Btw there are also some Ktlint errors now, but they could be automatically resolved by running `./gradlew formatKtlint`.

**Comment by TacoTheDank** [comment]
I'll try it.

**Comment by wb9688** [comment]
@TacoTheDank: No, you don't have to fix the warnings. Notice how it fails at `:runKtlint` now ;)

**Comment by TacoTheDank** [comment]
Ok, everything should NOW be good 😅

**Comment by TacoTheDank** [comment]
Also, running `./gradlew runCheckstyle` now just builds successfully without giving me any warnings at all. Very strange lol

**Comment by TacoTheDank** [comment]
I like my PRs being very clean, with no unnecessary after-changes. Sorry for all the force-pushing lol

**Comment by opusforlife2** [comment]
It's fine. I don't mind. You can force-push all you want. 🤭

**Comment by MD77MD** [comment]
https://github.com/TeamNewPipe/NewPipe/pull/3909#issuecomment-682191436  that sounds naughty 😁

**Comment by opusforlife2** [comment]
Oh no no no. All force-pushing is 100% consensual in this repo, no matter how scary the term sounds. 🤣

**Comment by Redirion** [comment]
> Code looks good, but will be merged after #2907 as it changes player code.

@TobiGr In my opinion this is good to merge now.

**Review by Stypox — APPROVED** [review]
Code looks good to me. Have you tested that everything works correctly on Android 4.4 with `SwitchPreferenceCompat`? (it should, but I remember strange shenanigans a while ago)

File: app/src/main/java/org/schabi/newpipe/player/MainPlayer.java [hunk 4]
```diff
@@ -307,9 +307,7 @@ void resetNotification() {
             setLockScreenThumbnail(builder);
         }
 
-        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.JELLY_BEAN) {
-            builder.setPriority(NotificationCompat.PRIORITY_MAX);
-        }
+        builder.setPriority(NotificationCompat.PRIORITY_MAX);
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/MainPlayer.java:310]: Heads up @friendlyanon

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java [hunk 5]
```diff
@@ -303,10 +302,6 @@ public static int getTossFlingVelocity(@NonNull final Context context) {
 
     @NonNull
     public static CaptionStyleCompat getCaptionStyle(@NonNull final Context context) {
-        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
-            return CaptionStyleCompat.DEFAULT;
-        }
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java:308]: Heads up @friendlyanon

File: app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java [hunk 6]
```diff
@@ -331,10 +326,6 @@ public static CaptionStyleCompat getCaptionStyle(@NonNull final Context context)
      * @return caption scaling
      */
     public static float getCaptionScale(@NonNull final Context context) {
-        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
-            return 1f;
-        }
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/helper/PlayerHelper.java:336]: Heads up @friendlyanon

File: app/src/main/java/org/schabi/newpipe/util/PermissionHelper.java [hunk 7]
```diff
@@ -26,10 +26,8 @@
     private PermissionHelper() { }
 
     public static boolean checkStoragePermissions(final Activity activity, final int requestCode) {
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            if (!checkReadStoragePermissions(activity, requestCode)) {
-                return false;
-            }
+        if (!checkReadStoragePermissions(activity, requestCode)) {
+            return false;
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/util/PermissionHelper.java:30]: @friendlyanon

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 8]
```diff
@@ -1105,7 +1105,7 @@ private void openBackgroundPlayer(final boolean append) {
             player.toggleFullscreen();
         }
 
-        if (!useExternalAudioPlayer && android.os.Build.VERSION.SDK_INT >= 16) {
+        if (!useExternalAudioPlayer) {
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:1108]: Heads up @friendlyanon

File: app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java [hunk 9]
```diff
@@ -1302,16 +1302,9 @@ private void prepareDescription(final Description description) {
...
-                            parsedDescription = Html.fromHtml(descriptionText, 0);
-                        } else {
-                            //noinspection deprecation
-                            parsedDescription = Html.fromHtml(descriptionText);
-                        }
-                        return parsedDescription;
-                    })
+                    .map((@NonNull String descriptionText) ->
+                            HtmlCompat.fromHtml(descriptionText,
+                                    HtmlCompat.FROM_HTML_MODE_LEGACY))
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/fragments/detail/VideoDetailFragment.java:1307]: @friendlyanon

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java [hunk 10]
```diff
@@ -476,11 +474,8 @@ public void onPlaying() {
         showAndAnimateControl(-1, true);
 
         playbackSeekBar.setEnabled(true);
-        // Bug on lower api, disabling and enabling the seekBar resets the thumb color -.-,
-        // so sets the color again
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            playbackSeekBar.getThumb().setColorFilter(Color.RED, PorterDuff.Mode.SRC_IN);
-        }
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:478]: @friendlyanon

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java [hunk 11]
```diff
@@ -456,11 +457,8 @@ public void onBlocked() {
         animateView(controlsRoot, false, DEFAULT_CONTROLS_DURATION);
 
         playbackSeekBar.setEnabled(false);
-        // Bug on lower api, disabling and enabling the seekBar resets the thumb color -.-,
-        // so sets the color again
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            playbackSeekBar.getThumb().setColorFilter(Color.RED, PorterDuff.Mode.SRC_IN);
-        }
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:461]: @friendlyanon

File: app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java [hunk 12]
```diff
@@ -210,18 +211,18 @@ public void initViews(final View view) {
 
         this.captionTextView = view.findViewById(R.id.captionTextView);
 
-        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
-            playbackSeekBar.getThumb().setColorFilter(Color.RED, PorterDuff.Mode.SRC_IN);
-        }
-        this.playbackSeekBar.getProgressDrawable().
-                setColorFilter(Color.RED, PorterDuff.Mode.MULTIPLY);
+        playbackSeekBar.getThumb()
+                .setColorFilter(new PorterDuffColorFilter(Color.RED, PorterDuff.Mode.SRC_IN));
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/player/VideoPlayer.java:215]: @friendlyanon

File: app/src/main/java/org/schabi/newpipe/streams/Mp4DashReader.java [hunk 13]
```diff
@@ -259,11 +259,7 @@ private String boxName(final Box ref) {
     }
 
     private String boxName(final int type) {
-        try {
-            return new String(ByteBuffer.allocate(4).putInt(type).array(), "UTF-8");
-        } catch (final UnsupportedEncodingException e) {
-            return "0x" + Integer.toHexString(type);
-        }
+        return new String(ByteBuffer.allocate(4).putInt(type).array(), StandardCharsets.UTF_8);
```
Comment (TobiGr) [inline app/src/main/java/org/schabi/newpipe/streams/Mp4DashReader.java:262]: @friendlyanon StandardCharsets was added with API 19

**Review by TobiGr — COMMENTED** [review]
Looks good to me.
At the reviewers: please tell @friendlyanon whenever there are changes which might cause problems with the legacy version to not introduce preventable regressions. @friendlyanon You basically need to take a look at 1da206363eee8acd4ab2a2d325f84cdcdcacf31e and `StandardCharset` changes.

**Comment by TacoTheDank** [comment]
@Stypox Unfortunately I do not have a KitKat device on which to test. :(

**Comment by Stypox** [comment]
Ok, I will test myself on an emulator then 😉

**Comment by Stypox** [comment]
I fixed a problem with the `TabAdapter` of `VideoDetailFragment`: for some reason the `BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT` had this issue: crash if enqueueing stream in the background and then clicking on it to open VideoDetailFragment: "Cannot setMaxLifecycle for Fragment not attached to FragmentManager". This can be solved by updating everything to [ViewPager2](https://developer.android.com/jetpack/androidx/releases/viewpager2) and removing `TabAdapter` altogether, but this is not relevant to this PR.
Everything else I checked worked fine, even on Android 4.4.

**Review by Stypox — APPROVED** [review]
Ready for merge in my opinion @wb9688. Thank you @TacoTheDank :-D

**Comment by TacoTheDank** [comment]
I agree with migrating to ViewPager2.

**Review by wb9688 — APPROVED** [review]

**Comment by Stypox** [comment]
@TacoTheDank thank you for your effort!
